### PR TITLE
feat!: checks the exact size of the node in HashNode

### LINKED ISSUE #153: ValidateNode does not check the exact size of the non-leaf nodes
# Problem
In an NMT, non-leaf nodes all have a particular size that is `2*namespace size + hash output size`. 
The `ValidateNode` should take into account this size and validate nodes' size accordingly. However, this is not the case in the current implementation. 

# Acceptance Criteria
Incorporate the exact size check to the `ValidateNode` method together with proper unit tests.

### PR DESCRIPTION
**Pull request by staheri14**
## Overview
Closes #153 (part of https://github.com/celestiaorg/nmt/issues/173)

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

### CODE REVIEW

File: hasher.go [hunk 1]
```diff
@@ -194,12 +194,12 @@ func (n *Hasher) MustHashLeaf(ndata []byte) []byte {
...
 // ValidateNodeFormat checks whether the supplied node conforms to the
-// namespaced hash format and returns an error if it does not. Specifically, it returns ErrInvalidNodeLen if the length of the node is less than the 2*namespace length which indicates it does not match the namespaced hash format.
+// namespaced hash format and  returns ErrInvalidNodeLen if not.
 func (n *Hasher) ValidateNodeFormat(node []byte) (err error) {
-	totalNamespaceLen := 2 * n.NamespaceLen
+	expectedNodeLen := n.Size()
 	nodeLen := len(node)
-	if nodeLen < int(totalNamespaceLen) {
-		return fmt.Errorf("%w: got: %v, want >= %v", ErrInvalidNodeLen, nodeLen, totalNamespaceLen)
+	if nodeLen < int(expectedNodeLen) {
```
Comment (evan-forbes) [inline hasher.go:201]: why do we only want to check `<` here instead of `!=`?

**Review by evan-forbes — COMMENTED** [review]
nice, LGTM, only had a single question

this change seems like it could be consensus breaking if some nodes use it and others don't, so we might want to rename this PR to `feat!...` to indicate such a change

**Comment by staheri14** [comment]
> this change seems like it could be consensus breaking if some nodes use it and others don't, so we might want to rename this PR to feat!... to indicate such a change

Right, updated the PR description.

**Review by evan-forbes — APPROVED** [review]
:+1: :+1:

File: hasher.go [hunk 2]
```diff
@@ -194,12 +194,12 @@ func (n *Hasher) MustHashLeaf(ndata []byte) []byte {
 }
 
 // ValidateNodeFormat checks whether the supplied node conforms to the
-// namespaced hash format and returns an error if it does not. Specifically, it returns ErrInvalidNodeLen if the length of the node is less than the 2*namespace length which indicates it does not match the namespaced hash format.
+// namespaced hash format and  returns ErrInvalidNodeLen if not.
```
Comment (rootulp) [inline hasher.go:197]:
```suggestion
// namespaced hash format and returns ErrInvalidNodeLen if not.
```

File: hasher.go [hunk 3]
```diff
@@ -194,12 +194,12 @@ func (n *Hasher) MustHashLeaf(ndata []byte) []byte {
...
 func (n *Hasher) ValidateNodeFormat(node []byte) (err error) {
-	totalNamespaceLen := 2 * n.NamespaceLen
+	expectedNodeLen := n.Size()
 	nodeLen := len(node)
-	if nodeLen < int(totalNamespaceLen) {
-		return fmt.Errorf("%w: got: %v, want >= %v", ErrInvalidNodeLen, nodeLen, totalNamespaceLen)
+	if nodeLen != expectedNodeLen {
+		return fmt.Errorf("%w: got: %v, want >= %v", ErrInvalidNodeLen, nodeLen, expectedNodeLen)
```
Comment (rootulp) [inline hasher.go:202]: [blocking] this error message is misleading b/c the conditional has been updated

```suggestion
		return fmt.Errorf("%w: got %v, want %v", ErrInvalidNodeLen, nodeLen, expectedNodeLen)
```

File: hasher_test.go [hunk 4]
```diff
@@ -277,6 +314,9 @@ func TestValidateSiblingsNamespaceOrder(t *testing.T) {
 }
 
 func TestValidateNodeFormat(t *testing.T) {
```
Comment (rootulp) [inline hasher_test.go:316]: [question] should a test case be added that is > the expected non-leaf node size?

**Review by rootulp — CHANGES_REQUESTED** [review]

File: hasher.go (hunk 2 above)
Comment (staheri14) [reply hasher.go]: Addressed https://github.com/celestiaorg/nmt/pull/175/commits/76b27b47eddfeddecd17532f544971458a979c33

File: hasher_test.go (hunk 4 above)
Comment (staheri14) [reply hasher_test.go:316]: Sure, done in https://github.com/celestiaorg/nmt/pull/175/commits/f5427e5c71271735866d20f093c25a5689aef6b2

**Review by rootulp — APPROVED** [review]

**Pull request merged**